drivers/overlay/ov_network.go

```diff
@@ -296,29 +294,39 @@ func (d *driver) RevokeExternalConnectivity(nid, eid string) error {
 	return nil
 }
 
-func (n *network) incEndpointCount() {
+func (n *network) joinSandbox(s *subnet, restore bool, incJoinCount bool) error {
+	networkOnce.Do(networkOnceInit)
+
 	n.Lock()
 	defer n.Unlock()
-	n.joinCnt++
-}
 
-func (n *network) joinSandbox(restore bool) error {
-	// If there is a race between two go routines here only one will win
-	// the other will wait.
-	n.once.Do(func() {
-		// save the error status of initSandbox in n.initErr so that
-		// all the racing go routines are able to know the status.
+	if !n.sboxInit {
 		n.initErr = n.initSandbox(restore)
-	})
+		n.sboxInit = true // we cannot recover from this error
+	}
 
-	return n.initErr
-}
+	if n.initErr != nil {
+		return fmt.Errorf("network sandbox join failed: %v", n.initErr)
+	}
 
-func (n *network) joinSubnetSandbox(s *subnet, restore bool) error {
-	s.once.Do(func() {
-		s.initErr = n.initSubnetSandbox(s, restore)
-	})
-	return s.initErr
+	subnetErr := s.initErr
```

Comment:
we can simply avoid using this subnetErr, and just leverage the s.initErr

Reply:
also for line 321

Reply:
Actually, there is a very specific reason why it isn't.  We want to return:
  - `s.initErr` *iff* it was previously set
  - otherwise we want to return the result of `n.initSubnetSandbox(s, restore)` *regardless* of whether that value gets stored in `s.initErr`.  So, on line 321 we do not want to just return `s.initErr`.  It may not have been set if there was a failure but we are not in a `restore` case.

This is the semantics from the https://github.com/docker/libnetwork/pull/1800 and I've preserved them.  The logic basically implies that if an error is recoverable, return it so it is flagged appropriately, but to make it persistent.  But if the error is not recoverable (restore case), make it persistent.

Reply:
do we expect that the error failure will be different? on multiple retry?